presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveParquetDereferencePushDown.java

```diff
@@ -0,0 +1,403 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.rule;
+
+import com.facebook.presto.common.type.RowType;
+import com.facebook.presto.expressions.DefaultRowExpressionTraversalVisitor;
+import com.facebook.presto.expressions.RowExpressionRewriter;
+import com.facebook.presto.expressions.RowExpressionTreeRewriter;
+import com.facebook.presto.hive.HiveColumnHandle;
+import com.facebook.presto.hive.HiveMetadata;
+import com.facebook.presto.hive.HiveTableHandle;
+import com.facebook.presto.hive.HiveTransactionManager;
+import com.facebook.presto.hive.HiveType;
+import com.facebook.presto.hive.NestedColumn;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorPlanOptimizer;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.VariableAllocator;
+import com.facebook.presto.spi.connector.ConnectorMetadata;
+import com.facebook.presto.spi.plan.Assignments;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.plan.PlanVisitor;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.ExpressionOptimizer;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionService;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.expressions.RowExpressionTreeRewriter.rewriteWith;
+import static com.facebook.presto.hive.HiveColumnHandle.ColumnType.NESTED_COLUMN;
+import static com.facebook.presto.hive.HiveSessionProperties.isParquetDereferencePushdownEnabled;
+import static com.facebook.presto.hive.HiveStorageFormat.PARQUET;
+import static com.facebook.presto.hive.HiveTableProperties.getHiveStorageFormat;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.DEREFERENCE;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+public class HiveParquetDereferencePushDown
```

Comment:
how about HiveDereferencePushDown? I think this is general for all formats, right?

Reply:
Currently this rule is applicable only for Parquet (there is check in the rule). Enabling it by default (without the storage type check) will cause issues if the reader is doesn't understand the HiveColumnHandle with NESTED_COLUMN type.